### PR TITLE
prometheus: Improve and modular for easy extendability

### PR DESCRIPTION
**Pull request by ztec**
prometheus: Improve and modular for easy extendability

I had a prometheus script running on my own for a while and thanks to @fiws I know have more data to export. I've reworked my script to includes @fiws initial version.

  - The script design is more modular to allow more than one switch and also to include other kind of metrics. (I have in mind wifi, BT, cloud, MQTT, settings, but I didn't implement any of them)
  - configurable prefix (I didn't set the same as @fiws but still want to share my version)
  - configuration of the url (Personally I use `metrics` because it is pretty standard in prometheus landscape, ~but I left status for bet backward compatibility~ Seems BC break is not so important for this one.
  - BREAKING_CHANGE: the `temperature_celsius` metrics is now called `switch_temperature_celsius` for coherence.
  - Added : I've added teh switch status (open/off/0 or closed/on/1) to monitor that as well
  - Added an explicit 200 status code. I have no idea if this is really required so
  - Added descriptions and usual HELP lines in the output to ease human reading of the metrics (is it is common in prometheus landscape)
  - BREAKING_CHANGE: `switch_power_watthours_total` is not called `switch_power_milliwatthours_total` to better represent actual unit

 Thank you, @fiws, for the missing peaces to assemble this.

I've included a Grafana example dashboard (I will try to upload it to grafana.com when I can)

Grafana dashboard showing switch information

### CODE REVIEW

File: prometheus.js [hunk 1]
```diff
@@ -12,39 +12,106 @@
...
+ */
+function generateMetricsForSwitch(string_id) {
+  const sw = Shelly.getComponentStatus(string_id);
+  const res = []
+  res.push(printPrometheusMetric("switch_power_watts", TYPE_GAUGE, [promLabel("switch", sw.id)], "Instant power consumption in watts", sw.apower))
+  res.push(printPrometheusMetric("switch_voltage_volts", TYPE_GAUGE, [promLabel("switch", sw.id)], "Instant voltage in volts", sw.voltage))
+  res.push(printPrometheusMetric("switch_current_amperes", TYPE_GAUGE, [promLabel("switch", sw.id)], "Instant current in amperes", sw.current))
+  res.push(printPrometheusMetric("switch_temperature_celsius", TYPE_GAUGE, [promLabel("switch", sw.id)], "temperature of the plug in celsius", sw.temperature.tC))
+  res.push(printPrometheusMetric("switch_power_watthours_total", TYPE_COUNTER, [promLabel("switch", sw.id)], "accumulated energy consumed in watts hours", sw.aenergy.total))
+  res.push(printPrometheusMetric("switch_output", TYPE_GAUGE, [promLabel("switch", sw.id)], "is switch (1)on or (0)off", sw.output ? 1 : 0))
```
[the remark this answers is not in the document]
Comment (ztec) [reply prometheus.js]: Yup, did it for both and removed the array completely. Thanks

**Comment by ztec** [comment]
The breaking changes will be bigger on my side :-D I had one running for month. But I don't really care. 

I can rename the endpoint to `metrics` then.

**Comment by ztec** [comment]
I had to fix one of the metrics and now `switch_power_watthours_total` is caleld `switch_power_milliwatthours_total` to better reflect the unit it really represent.

**Comment by ztec** [comment]
I'm a bit confused with the total wat hour counters. On my side it give strange numbers when used with a `rate` 
I'm not sure what is reall ythe unit there (Wh or mWh ?)

**Comment by ztec** [comment]
Nevermind, I had previous `mWh` set as unit in grafana from my previous script version that was actually exporting milliWh

**Comment by ztec** [comment]
Also, does your `ram_size` change over time ?

**Comment by ztec** [comment]
after my last install, numbers seems pretty stable (no single change) for few hours. I'm curious as why it was changing a lot before :thinking:  
image

**Comment by Mircho** [comment]
Thank you! Providing the additional blueprint for a dashboard is nice touch. We are working on packaging scripts + resources as apps, but that will take some time.

**Comment by fiws** [comment]
nice. @Mircho could you shed some light on the `ram_size` field? Is it a bug that it fluctuates? Or is the naming just incorrect?

**Comment by Mircho** [comment]
Digging through the source code ram_size is heap size. Not very precise wording from our side.

**Comment by ztec** [comment]
ah, interesting. Thanks for the details.

**Pull request merged**